### PR TITLE
devenv: 3.10 for pypi, factor out venvs_root

### PR DESCRIPTION
**Pull request by joshuarli**
3.10 required for internal pypi

factors out venvs_root which is not a great idea, instead venvs are now repo local (doesn't affect sentry either way cause `path` can be configured to be `.venv`)

### CODE REVIEW

**Review by asottile-sentry — COMMENTED** [review]
we should probably be testing in 3.10 as well if we're going to claim support for it here

**Comment by joshuarli** [comment]
> we should probably be testing in 3.10 as well if we're going to claim support for it here

lol adding 3.10 to unittests uncovered some disk pollution, will fix

not sure if its really worth 2xing integration tests though

**Comment by asottile-sentry** [comment]
> > we should probably be testing in 3.10 as well if we're going to claim support for it here
> 
> lol adding 3.10 to unittests uncovered some disk pollution, will fix
> 
> not sure if its really worth 2xing integration tests though

could matrix them maybe -- though they're like seconds to finish (and this uncovered a bug!)

will want to add 3.12 anyway since I want to get that going

**Comment by joshuarli** [comment]
ok so this boils down to `os.environ["HOME"] = f"{tmp_path}"` in a test before importing `devenv` isn't reliable still

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -18,6 +18,7 @@ repos:
     rev: v1.6.1
     hooks:
       - id: mypy
+        exclude: '^tests/'
```
Comment (asottile-sentry) [inline .pre-commit-config.yaml:21]: tests should always be checked for typing problems

File: requirements-dev.in [hunk 2]
```diff
@@ -2,3 +2,4 @@ mypy
 pytest
 pre-commit
 black
+pyfakefs
```
Comment (asottile-sentry) [inline requirements-dev.in:5]: I would not recommend this package -- if you're reaching for this it indicates you're not properly fixturing.  use `tmp_path` pytest fixture for your paths and you should be good

**Review by asottile-sentry — CHANGES_REQUESTED** [review]

**Comment by asottile-sentry** [comment]
> going down the path of how do i get a test's tmp_path to take effect by reimporting devenv seems hacky

this is a good indication that the global variables referring to paths are a design problem and should be refactored to not depend on global state

**Comment by joshuarli** [comment]
> > going down the path of how do i get a test's tmp_path to take effect by reimporting devenv seems hacky
> 
> this is a good indication that the global variables referring to paths are a design problem and should be refactored to not depend on global state

i can't think of a better design... can you? user's home path is pretty fundamental and it'd be a pita to move it and everything based on it to function local... is pointing HOME to tmp_path that bad? because that's all i need to figure out how to do

**Comment by asottile-sentry** [comment]
> > > going down the path of how do i get a test's tmp_path to take effect by reimporting devenv seems hacky
> > 
> > 
> > this is a good indication that the global variables referring to paths are a design problem and should be refactored to not depend on global state
> 
> i can't think of a better design... can you? user's home path is pretty fundamental and it'd be a pita to move it and everything based on it to function local... is pointing HOME to tmp_path that bad even if it were neatly possible under pytest (it isn't)?

the problem is nothing is testable if it's depending on the globals -- you might be able to `mock.patch.object` each of them but it's going to be a lot cleaner if they're not global

**Review by asottile-sentry — APPROVED** [review]

**Review by asottile-sentry — APPROVED** [review]

**Pull request merged**